lib/trino-parquet/src/main/java/io/trino/parquet/reader/flat/FilteredRowRangesIterator.java

```diff
@@ -0,0 +1,193 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.parquet.reader.flat;
+
+import io.trino.parquet.reader.FilteredRowRanges;
+
+import java.util.Iterator;
+import java.util.OptionalLong;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static io.trino.parquet.reader.FilteredRowRanges.RowRange;
+import static java.lang.Math.min;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * When filtering using column indexes we might skip reading some pages for different columns. Because the rows are
+ * not aligned between the pages of the different columns it might be required to skip some values. The values (and the
+ * related rl and dl) are skipped based on the iterator of the required row indexes and the first row index of each
+ * page.
+ * For example:
+ *
+ * <pre>
+ * rows   col1   col2   col3
+ *      ┌──────┬──────┬──────┐
+ *   0  │  p0  │      │      │
+ *      ╞══════╡  p0  │  p0  │
+ *  20  │ p1(X)│------│------│
+ *      ╞══════╪══════╡      │
+ *  40  │ p2(X)│      │------│
+ *      ╞══════╡ p1(X)╞══════╡
+ *  60  │ p3(X)│      │------│
+ *      ╞══════╪══════╡      │
+ *  80  │  p4  │      │  p1  │
+ *      ╞══════╡  p2  │      │
+ * 100  │  p5  │      │      │
+ *      └──────┴──────┴──────┘
+ * </pre>
+ *
+ * The pages 1, 2, 3 in col1 are skipped, so we have to skip the rows [20, 79]. Because page 1 in col2 contains values
+ * only for the rows [40, 79] we skip this entire page as well. To synchronize the row reading we have to skip the
+ * values (and the related rl and dl) for the rows [20, 39] in the end of the page 0 for col2. Similarly, we have to
+ * skip values while reading page0 and page1 for col3.
+ */
+public class FilteredRowRangesIterator
+        implements RowRangesIterator
+{
+    // Iterator over all row ranges, only not-null if column index is present
+    private final Iterator<RowRange> rowRangeIterator;
+
+    // The current row range
+    private RowRange currentRange;
+
+    private long pageFirstRowIndex = Long.MAX_VALUE;
+    private int pageValuesConsumed;
+
+    public FilteredRowRangesIterator(FilteredRowRanges rowRanges)
+    {
+        requireNonNull(rowRanges, "rowRanges is null");
+        this.rowRangeIterator = rowRanges.rowRanges().iterator();
+        // If rowRanges is empty then all pages would be eliminated by filterOffsetIndex
```

Comment:
Make `FilteredRowRangesIterator` agnostic about whether `rowRanges` is empty and then this comment and additional explanation would be unnecessary. It also reduces behavior coupling with the caller.

Reply:
Adding empty rowRanges checks to this class would unnecessarily complicate the code.
I could add a `NoneRowRangesIterator` implementation in `RowRangesIterator#createRowRangesIterator` for the empty rowRanges case, however that would be dead code.
I've updated the comment here to explain that we don't handle empty rowRanges because in that case, no page needs to be read, so a `ColumnReader` shouldn't get created and `FilteredRowRangesIterator` shouldn't get instantiated. If we reach here with empty rowRanges, then it's a bug which would be caught by the explicit `checkArgument` here.